common/clock/ratelimiter_test.go

```diff
@@ -0,0 +1,434 @@
+)
+
+func TestAgainstRealRatelimit(t *testing.T) {
+	// The mock ratelimiter should behave the same as a real ratelimiter in all cases.
+	//
+	// So fuzz test it: throw random calls at non-overlapping times that are a bit
+	// spaced out (to prevent noise due to busy CPUs) and make sure all impls agree.
+	//
+	// If a test fails, please verify by hand with the failing seed.
+	// Enabling debug logs can help show what's happening and what *should* be happening,
+	// though there is a chance it's just due to CPU contention.
+	const (
+		maxBurst = 10
+
+		// amount of time between each conceptual "tick" of the test's clocks, both real and fake.
+		// sadly even 10ms has some excessive latency a few % of the time, particularly during the Wait check (10ms just to start a goroutine!).
+		granularity = 20 * time.Millisecond
+		// number of [granularity] events to trigger per round,
+		// and also the number of [granularity] periods before a token is added.
+		//
+		// each test takes at least granularity*events*rounds, so keep it kinda small.
+		events = 10
+		// number of rounds of "N events" to try
+		rounds = 2
+
+		// keep running tests until this amount of time has passed, or a failure has occurred.
+		// this could be a static count of tests to run, but a target duration
+		// is a bit less fiddly.  "many attempts" is the goal, not any specific number.
+		//
+		// alternatively, tests have a deadline, this could run until near that deadline.
+		// but right now that's not fine-tuned per package, so it's quite long.
+		testDuration = 4 * time.Second // mostly stays under 5s, feels reasonable
+	)
+
+	// normally, this produces a ton of text that's just noise during successful runs.
+	// so for bulk verbose modes, like `make test` does, skip the debug logs unless force-enabled.
+	//
+	// regrettably this makes failures less useful when verbose, but the
+	// failure logs do at least inform about this behavior.
+	lowVerbosity := testing.Verbose()
+
+	t.Cleanup(func() {
+		if t.Failed() {
+			/*
+				Is this test becoming too slow or too noisy?
+
+				The easiest "likely to actually work" fix is probably going to
+				require detecting excessive lag, and retrying instead of failing.
+
+				That could miss some real flaws if they are racy by nature, but
+				seems like it might be good enough.
+			*/
+			t.Logf("---- CAUTION ----")
+			t.Logf("these tests are randomized by design, a random failure may be a real flaw!")
+			t.Logf("please replay with the failing seed and check detailed output to see what the behavior should be.")
+			if lowVerbosity {
+				t.Logf("try setting `lowVerbosity` to false to verbosely log these tests at all times.")
+			}
+			t.Logf("")
+			t.Logf("if you are intentionally making changes, be sure to run a few hundred rounds to make sure your changes are stable")
+			t.Logf("---- CAUTION ----")
+		}
+	})
+
+	debug := func(t *testing.T, format string, args ...interface{}) {
+		t.Logf(format, args...)
+	}
+	if lowVerbosity {
```

Comment:
`testing.Verbose() == true` means `test run -v ...`. Why is this called lowVerbosity?

Reply:
ah, I used to have a comment about that here.  evidently I removed it somewhere along the way.

basically: because the `debug(...)` stuff is very high verbosity, and it's really not useful _in bulk_, like you get with `go test -v` (or `make test` which runs ALL tests with `-v`).  it's like 50-100 lines per test (times N tests), when most of the useful information is covered in ~5-10 per test.

so I reversed the verbosity.  that way if it's not verbose (like in CI), you get fine-grained logs only when it fails, and silence otherwise.